src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -1811,52 +1811,58 @@ class MatrixTable(val hc: HailContext, val metadata: VSMMetadata,
   def reorderSamples(newIds: Array[Annotation]): MatrixTable = {
     requireUniqueSamples("reorder_samples")
 
-    val oldOrder = sampleIds.zipWithIndex.toMap
-    val newOrder = newIds.zipWithIndex.toMap
-
-    val newIndices = new Array[Int](nSamples)
-    val missingSamples = mutable.Set[Annotation]()
-    val notInDataset = mutable.Set[Annotation]()
-
-    oldOrder.outerJoin(newOrder).foreach { case (s, (oldIdx, newIdx)) =>
-      ((oldIdx, newIdx): @unchecked) match {
-        case (Some(i), Some(j)) => newIndices(i) = j
-        case (Some(i), None) => missingSamples += s
-        case (None, Some(j)) => notInDataset += s
-      }
-    }
+    val sampleSet = sampleIds.toSet
+    val newSampleSet = newIds.toSet
 
+    val missingSamples = sampleSet -- newSampleSet
     if (missingSamples.nonEmpty)
       fatal(s"Found ${ missingSamples.size } ${ plural(missingSamples.size, "sample ID") } in dataset that are not in new ordering:\n  " +
         s"@1", missingSamples.truncatable("\n  "))
 
+    val notInDataset = newSampleSet -- sampleSet
     if (notInDataset.nonEmpty)
       fatal(s"Found ${ notInDataset.size } ${ plural(notInDataset.size, "sample ID") } in new ordering that are not in dataset:\n  " +
         s"@1", notInDataset.truncatable("\n  "))
 
-    val newAnnotations = new Array[Annotation](nSamples)
-    sampleAnnotations.zipWithIndex.foreach { case (sa, idx) =>
-      newAnnotations(newIndices(idx)) = sa
+    val oldIndex = sampleIds.zipWithIndex.toMap
+    val newToOld = newIds.map(oldIndex)
+
+    val newAnnotations = Array.tabulate(nSamples) { i =>
+      sampleAnnotations(newToOld(i))
     }
 
-    val nSamplesLocal = nSamples
+    val localNSamples = nSamples
+    val localRowType = rowType
+    val localGType = genotypeSignature
 
-    val reorderedRdd = rdd.mapPartitions({ it =>
-      it.map { case (v, (va, gs)) =>
-        val reorderedGs = new Array[Annotation](nSamplesLocal)
-        val gsIt = gs.iterator
+    val reorderedRDD2 = rdd2.mapPartitionsPreservesPartitioning(rdd2.typ) { it =>
+      val rvb = new RegionValueBuilder()
+      val rv2 = RegionValue()
 
+      it.map { rv =>
+        rvb.set(rv.region)
+        rvb.start(localRowType)
+        rvb.startStruct()
+        rvb.addField(localRowType, rv, 0) // pk
+        rvb.addField(localRowType, rv, 1) // v
+        rvb.addField(localRowType, rv, 2) // va
+        rvb.startArray(localNSamples) // gs
+        val ur = new UnsafeRow(localRowType, rv)
```

Comment:
can't you move this outside the `map` and just set it to the region?

Reply:
How would you know the offset?

Reply:
I think he means the allocation of ur, and then reuse it in the map.  Yes, I can do that.

Reply:
it looks like UnsafeRow has a `set` method for this case. Wouldn't it work here? Why not if it won't?

Reply:
Oh, got it